### ISSUE TITLE
False positive when using `type` in overloaded generic function

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#3548]
**Describe the bug**
I'm getting a static type checking error when I'm trying to create an overloaded function with `type[T]` annotation in some of the overload signatures.

**To Reproduce**

```python
from typing import TypeVar, overload

T = TypeVar("T")

@overload
def f(arg: type[T]) -> None:
    ...

@overload
def f(arg: str) -> None:
    ...

def f(arg: object) -> None:
    pass
```

The above code produces the following error:

```
error: Overloaded implementation is not consistent with signature of overload 1
    Type "(arg: object) -> None" cannot be assigned to type "(arg: Type[T@f]) -> None"
      Parameter 1: type "Type[T@f]" cannot be assigned to type "object"
        Type "object" cannot be assigned to type "Type[T@f]"
```

I wonder why is Pyright trying to assign type `object` to type `type[T]` and not the other way around. I also made a little investigation and found out that the error disappears if I use nonparameterized annotation (just `arg: type`) or if I replace `type` with some other generic type (`list[T]`, etc).

Pyright version used: `1.1.252`.

**Comment by erictraut** [comment]
Pyright is working as intended here. I don't consider this a false positive.

First, your use of `T` in the first overload is an incorrect usage of a type variable. There's no reason to include a TypeVar in this case because the TypeVar appears only once, and it has no bound. TypeVars are mean to be constraints for a constraint solver, but with a single TypeVar instance, there is no constraint, so it is effectively `type[Any]` or just `type`.

If you do include a TypeVar `T` in the first overload, the same should appear in the implementation. But since there's no reason to include `T` in the first overload, the correct fix is to remove it.

```python
from typing import overload

@overload
def f(arg: type) -> None:
    ...

@overload
def f(arg: str) -> None:
    ...

def f(arg: object) -> None:
    pass
```

**Comment by LeeeeT** [comment]
Maybe I gave a bad example. My original code looks something like this:

```python
@overload
def parse(to_type: type[T]) -> T:
    ...

@overload
def parse(to_type: object) -> Any:
    ...

def parse(to_type: object) -> Any:
    raise NotImplementedError

```

I want to create a function that takes some type and returns an instance of that type. But the function can get `int | str` or even a forward ref `"MyType"`, for example. And this is assumed to be a valid value. But it's currently impossible to annotate the return type in this case, so I'm using `Any` in the second signature and getting the same error.

```
error: Overloaded implementation is not consistent with signature of overload 1
    Type "(to_type: object) -> Any" cannot be assigned to type "(to_type: Type[T@parse]) -> T@parse"
      Parameter 1: type "Type[T@parse]" cannot be assigned to type "object"
        Type "object" cannot be assigned to type "Type[T@parse]"
```

**Comment by erictraut** [comment]
Then you should annotate your implementation as:
```python
def parse(to_type: type | object) -> Any:
    raise NotImplementedError
```

or
```python
def parse(to_type: Any) -> Any:
    raise NotImplementedError
```

**Comment by LeeeeT** [comment]
Finally found a good solution:

```python
@overload
def parse(to_type: type[T]) -> T:
    ...

@overload
def parse(to_type: object) -> Any:
    ...

def parse(to_type: type[T] | object) -> T | Any:
    raise NotImplementedError
```

But it's still unclear why there're no errors if I replace `type[T]` with `list[T]` in my [very first example](https://github.com/microsoft/pyright/issues/3548#issue-1262356176). As you said, such usage of a type variable is incorrect.
